WordPress/gutenberg
# CustomSelectControl: support generic props type

## What?


Fixes #63984

## Why?


It would be helpful to infer the option type

## How?


- Making the `CustomSelectProps` a generic type
- Exporting `CustomSelectOption` and `CustomSelectChangeObject` type

## Testing Instructions

Using the following snippet to see if the `selectedItem` type is inferred correctly:

```tsx
import { CustomSelectControl } from '@wordpress/components';
import type { CustomSelectOption } from '@wordpress/components/build-types/custom-select-control/types';

type MyOption = CustomSelectOption & {
  foo: string;
};

const options: MyOption[] = [
  {
    key: 'none',
    name: 'None',
    foo: 'bar',
  },
  {
    key: 'foo',
    name: 'Foo',
    foo: 'bar',
  },
];

const render = () => (
  <CustomSelectControl
    label="foo"
    onChange={({ selectedItem }) => {
      console.log(selectedItem.foo);
    }}
    options={options}
    value={options[0]}
  />
);
```

### Testing Instructions for Keyboard


N/A

## Screenshots or screencast 

**Before**

<img width="760" alt="image" src="https://github.com/user-attachments/assets/241e95bb-5bf8-4c3e-aecd-ff46cfb03944">

**After**

<img width="533" alt="image" src="https://github.com/user-attachments/assets/6d9982f0-562c-4b17-ab06-903d9aef131c">
